### PR TITLE
Emep mulitple years

### PR DESCRIPTION
**Pull request by dulte**
Reads and combines emep model data for multiple years

### CODE REVIEW

**Comment by jgliss** [comment]
Cool, make sure to modify affected tests and add new ones for the new feature(s).

**Comment by dulte** [comment]
Thanks! There are still many things to do, and I don't think it actually works yet. But I made the draft just to have done it :)

File: pyaerocom/colocation_auto.py [hunk 1]
```diff
@@ -1008,8 +1008,8 @@ def _instantiate_gridded_reader(self, what):
             data_dir = self.obs_data_dir
         reader_class = self._get_gridded_reader_class(what=what)
         reader = reader_class(data_id=data_id, data_dir=data_dir)
-        if isinstance(reader, ReadMscwCtm) and hasattr(self, 'filepath'):
-            reader.filepath = self.filepath
+        # if isinstance(reader, ReadMscwCtm) and hasattr(self, 'filepath'):
+        #     reader.filepath = self.filepath
```
Comment (avaldebe) [inline pyaerocom/colocation_auto.py:1012]: If this code is no longer needed, remove it.
Don't keep old code arround.

File: pyaerocom/io/read_mscw_ctm.py [hunk 2]
```diff
@@ -201,27 +283,52 @@ def filename(self, val):
...
         """
         Path to data file
         """
-        if self.data_dir is None:
-            raise AttributeError('need data_dir to be set in data')
-        return os.path.join(self.data_dir, self.filename)
+        if self.data_dir is None and self._filepaths is None:
+            raise AttributeError('data_dir or filepaths needs to be set before accessing')
+        return self._filepath
+        #return os.path.join(self.data_dir, self.filename)
```
Comment (avaldebe) [inline pyaerocom/io/read_mscw_ctm.py:301]: remove old code

File: pyaerocom/io/read_mscw_ctm.py [hunk 3]
```diff
@@ -1,7 +1,11 @@
+from posixpath import join
```
Comment (avaldebe) [inline pyaerocom/io/read_mscw_ctm.py:1]: I do not see this import being used

File: pyaerocom/io/read_mscw_ctm.py [hunk 4]
```diff
@@ -201,27 +283,52 @@ def filename(self, val):
...
+        #return os.path.join(self.data_dir, self.filename)
+
+    @filepaths.setter
+    def filepaths(self, value):
+        if not isinstance(value, list):
+            raise ValueError('needs to be list of strings')
+        self._filepaths = value
+        # ddir, fname = os.path.split(value)
+        # self.data_dir = ddir
+        # self.filename = fname
```
Comment (avaldebe) [inline pyaerocom/io/read_mscw_ctm.py:330]: remove old code

File: pyaerocom/io/read_mscw_ctm.py [hunk 5]
```diff
@@ -524,7 +649,7 @@ def read_var(self, var_name, ts_type=None, **kwargs):
         # At this point a GriddedData object with name gridded should exist
 
         gridded.metadata['data_id'] = self.data_id
-        gridded.metadata['from_files'] = [self.filepath]
+        gridded.metadata['from_files'] = self.filepaths # [self.filepaths]
```
Comment (avaldebe) [inline pyaerocom/io/read_mscw_ctm.py:652]: remove old code

File: tests/io/test_read_mscw_ctm.py [hunk 6]
```diff
@@ -65,51 +66,52 @@
...
-    (None,None,'blaaaa',{},pytest.raises(FileNotFoundError) ),
-    (None,None,EMEP_DIR+'/Base_month.nc',{},pytest.raises(ValueError)),
+    # (None,None,None, {'_data_dir' : None,'_filename' : 'Base_day.nc',
+    #                   '_filedata': None, '_file_mask' : None,
+    #                   '_files'    : None},
+    #     does_not_raise_exception()),
+    # ('blaaa',None,None,{},pytest.raises(FileNotFoundError)),
+    # (EMEP_DIR,None,None,{},pytest.raises(ValueError)),
+    (None,'blaaaa',{},pytest.raises(FileNotFoundError) ),
+    # (None,None,EMEP_DIR+'/Base_month.nc',{},pytest.raises(ValueError)),
```
Comment (avaldebe) [inline tests/io/test_read_mscw_ctm.py:83]: remove old code

File: tests/io/test_read_mscw_ctm.py [hunk 7]
```diff
@@ -65,51 +66,52 @@
...
-    (EMEP_DIR,None,None,{},pytest.raises(ValueError)),
-    (None,None,'blaaaa',{},pytest.raises(FileNotFoundError) ),
-    (None,None,EMEP_DIR+'/Base_month.nc',{},pytest.raises(ValueError)),
+    # (None,None,None, {'_data_dir' : None,'_filename' : 'Base_day.nc',
+    #                   '_filedata': None, '_file_mask' : None,
+    #                   '_files'    : None},
+    #     does_not_raise_exception()),
+    # ('blaaa',None,None,{},pytest.raises(FileNotFoundError)),
+    # (EMEP_DIR,None,None,{},pytest.raises(ValueError)),
+    (None,'blaaaa',{},pytest.raises(FileNotFoundError) ),
```
Comment (avaldebe) [inline tests/io/test_read_mscw_ctm.py:82]: the FileNotFoundError is a different test

File: tests/io/test_read_mscw_ctm.py [hunk 8]
```diff
@@ -65,51 +66,52 @@
...
+        reader = ReadMscwCtm(data_id, data_dir)
         for key, val in check.items():
             _val = getattr(reader, key)
             assert val == _val
 
 @pytest.mark.parametrize('value, raises', [
     (EMEP_DIR, does_not_raise_exception()),
-    (None, pytest.raises(TypeError)),
+    (None, pytest.raises(ValueError)),
     ('', pytest.raises(FileNotFoundError))
```
Comment (avaldebe) [inline tests/io/test_read_mscw_ctm.py:97]: ValueError and FileNotFoundError are a different test

File: tests/io/test_read_mscw_ctm.py [hunk 9]
```diff
@@ -65,51 +66,52 @@
...
+    reader = ReadMscwCtm(value)
     with raises:
         reader.data_dir = value
         assert os.path.samefile(reader.data_dir, value)
 
 @pytest.mark.parametrize('value, raises', [
-    ('', does_not_raise_exception()),
+    (EMEP_DIR, does_not_raise_exception()),
+    ('', pytest.raises(FileNotFoundError)),
     (None, pytest.raises(ValueError)),
```
Comment (avaldebe) [inline tests/io/test_read_mscw_ctm.py:108]: FileNotFoundError and ValueError are a different test

File: tests/io/test_read_mscw_ctm.py [hunk 10]
```diff
@@ -65,51 +66,52 @@
...
-def test_ReadMscwCtm_filename(value, raises):
+def test_ReadMscwCtm_data_dir(value, raises):
     reader = ReadMscwCtm()
     with raises:
-        reader.filename = value
-        assert reader.filename == value
+        reader.data_dir = value
+        assert reader.data_dir == value
 
 @pytest.mark.parametrize('value, raises, fmask, num_matches', [
```
Comment (avaldebe) [inline tests/io/test_read_mscw_ctm.py:116]: none of the tests raises an Exception, remove the raises param

File: tests/io/test_read_mscw_ctm.py [hunk 11]
```diff
@@ -185,29 +187,29 @@ def test_ReadMscwCtm_data(path_emep):
 
 @data_unavail
 def test_ReadMscwCtm_directory(path_emep):
-    data_dir = path_emep['data_dir']
+    data_dir =EMEP_DIR # path_emep['data_dir']
```
Comment (avaldebe) [inline tests/io/test_read_mscw_ctm.py:190]: remove old code

File: tests/io/test_read_mscw_ctm.py [hunk 12]
```diff
@@ -185,29 +187,29 @@ def test_ReadMscwCtm_data(path_emep):
...
+#      ['hourly','monthly','daily','yearly'], does_not_raise_exception())
+# ])
+# def test_ReadMscwCtm_ts_types(files, ts_types, raises, tmpdir):
+#     ddir = None
+#     for filename in files:
+#         open(os.path.join(tmpdir, filename), 'w').close()
+#         ddir = str(tmpdir)
+#         with raises:
+#             r = ReadMscwCtm(data_dir=ddir)
+#             assert sorted(r.ts_types) == sorted(ts_types)
```
Comment (avaldebe) [inline tests/io/test_read_mscw_ctm.py:212]: remove old code

File: tests/io/test_read_mscw_ctm.py [hunk 13]
```diff
@@ -233,9 +235,9 @@ def test_ReadMscwCtm_filename_from_ts_type(reader, filename, ts_type, raises):
         assert reader.filename_from_ts_type(ts_type) == filename
 
 def test_ReadMscwCtm_years_avail(path_emep):
-    data_dir = path_emep['data_dir']
+    data_dir = EMEP_DIR#path_emep['data_dir']
```
Comment (avaldebe) [inline tests/io/test_read_mscw_ctm.py:238]: remove old code

File: tests/io/test_read_mscw_ctm.py [hunk 14]
```diff
@@ -246,9 +248,9 @@ def test_ReadMscwCtm_open_file(path_emep):
     reader = ReadMscwCtm()
     with pytest.raises(AttributeError):
         reader.open_file()
-    reader.data_dir = path_emep['data_dir']
+    reader.data_dir = EMEP_DIR#path_emep['data_dir']
```
Comment (avaldebe) [inline tests/io/test_read_mscw_ctm.py:251]: remove old code

File: tests/io/test_read_mscw_ctm.py [hunk 15]
```diff
@@ -371,3 +379,164 @@ def test_read_emep_dummy_data(tmpdir,file_vars_and_units,freq,add_read,
             for var, mean in chk_mean.items():
                 np.testing.assert_allclose(objs[var].cube.data.mean(), mean,
                                            atol=0.1)
+
+
+@pytest.mark.parametrize('yr, test_yrs, freq,raises', [
+     ("", [2013, 2015, 2017, 2018, 2019], 'day',does_not_raise_exception()),
+     ("", [2013, 2015, 2017, 2018, 2019, 2012], 'month',pytest.raises(ValueError)),
+     ("", [2013, 2016, 2017], 'day',pytest.raises(ValueError)),
```
Comment (avaldebe) [inline tests/io/test_read_mscw_ctm.py:387]: ValueError is a different test

File: tests/io/test_read_mscw_ctm.py [hunk 16]
```diff
@@ -371,3 +379,164 @@ def test_read_emep_dummy_data(tmpdir,file_vars_and_units,freq,add_read,
...
+
+
+
+
+@pytest.mark.parametrize('freq, ts_type, raises', [
+     ('day', 'day',does_not_raise_exception()),
+     ('month','month',does_not_raise_exception()),
+     ('month', 'minute',pytest.raises(ValueError)),
+     ('day', 'daily',pytest.raises(ValueError)),
+     ('month', 'LF_month',pytest.raises(ValueError)),
```
Comment (avaldebe) [inline tests/io/test_read_mscw_ctm.py:427]: ValueError is a different test

File: tests/io/test_read_mscw_ctm.py [hunk 17]
```diff
@@ -371,3 +379,164 @@ def test_read_emep_dummy_data(tmpdir,file_vars_and_units,freq,add_read,
...
+        filepaths = reader.filepaths
+        wrong_file = os.path.join(os.path.split(filepaths[0])[0], f"Base_{ts_type}.nc")
+        filepaths[0] = wrong_file
+        new_yrs = reader._get_yrs_from_filepaths()
+
+        
+        cleaned_paths = reader._clean_filepaths(filepaths, new_yrs, tst)
+
+@pytest.mark.parametrize('extra_year, raises', [
+     (True,pytest.raises(ValueError)),
```
Comment (avaldebe) [inline tests/io/test_read_mscw_ctm.py:446]: ValueError is a different test

File: tests/io/test_read_mscw_ctm.py [hunk 18]
```diff
@@ -371,3 +379,164 @@ def test_read_emep_dummy_data(tmpdir,file_vars_and_units,freq,add_read,
...
+            new_yrs = reader._get_yrs_from_filepaths()
+        else:
+            new_yrs = reader._get_yrs_from_filepaths()
+
+        cleaned_paths = reader._clean_filepaths(filepaths, new_yrs, tst)
+        
+
+@pytest.mark.parametrize('yr, freq ,raises', [
+     ("", 'day',does_not_raise_exception()),
+     ("", 'hour',pytest.raises(ValueError)),
```
Comment (avaldebe) [inline tests/io/test_read_mscw_ctm.py:471]: ValueError is a different test

File: tests/io/test_read_mscw_ctm.py [hunk 19]
```diff
@@ -371,3 +379,164 @@ def test_read_emep_dummy_data(tmpdir,file_vars_and_units,freq,add_read,
...
+
+        data = reader.read_var("prmm", ts_type=tst)
+        assert data.ts_type == tst
+
+
+@pytest.mark.parametrize('yr, freq ,raises', [
+     ("", 'day',does_not_raise_exception()),
+     ("2017", 'month', does_not_raise_exception()),
+     ("2019", 'hour',does_not_raise_exception()),
+])
```
Comment (avaldebe) [inline tests/io/test_read_mscw_ctm.py:500]: no exceptions here, remove raises param

File: tests/io/test_read_mscw_ctm.py [hunk 20]
```diff
@@ -371,3 +379,164 @@ def test_read_emep_dummy_data(tmpdir,file_vars_and_units,freq,add_read,
...
+            assert len(reader.filepaths) == 5
+        else:
+            assert len(reader.filepaths) == 1
+
+@pytest.mark.parametrize('yr, freq ,raises', [
+     ("", ['day'],does_not_raise_exception()),
+     ("", ['day', 'hour'],does_not_raise_exception()),
+     ("2017", ['month'], does_not_raise_exception()),
+     ("2019", ['hour','day','month'],does_not_raise_exception()),
+])
```
Comment (avaldebe) [inline tests/io/test_read_mscw_ctm.py:528]: no exceptions here, remove raises param

File: tests/test_colocation_auto.py [hunk 21]
```diff
@@ -332,7 +332,7 @@ def test_colocator_instantiate_gridded_reader(path_emep):
     col.model_id = model_id
     r = col._instantiate_gridded_reader(what='model')
     assert isinstance(r, ReadMscwCtm)
-    assert r.filepath == col.filepath
+    #assert r.filepaths == col.filepath
```
Comment (avaldebe) [inline tests/test_colocation_auto.py:335]: remove old code

**Review by avaldebe — CHANGES_REQUESTED** [review]
please remove old/commented code and split tests with/without exceptions.
I can help you with the tests, just met me know.

**Comment by dulte** [comment]
@avaldebe, I've split all the rest for the EMEP reader now. I left one test with both does_not_raise_exception and a ValueError, since the purpose of the test was to test whether we get an error or not.

**Comment by dulte** [comment]
The other three test files are not written for this reader, but have only small changes to make them work with the new error. So if we are going to split all the test in these file, I suggest doing that in an other PR

File: pyaerocom/io/_read_mscw_ctm_helpers.py [hunk 22]
```diff
@@ -188,11 +188,11 @@ def calc_vmrox(concno2, vmro3):
...
-    o3mulfac = get_unit_conversion_fac(1, vmro3.attrs['units'], 'nmole mole-1')
-    if o3mulfac != 1:
-        vmro3 = vmro3.copy(deep=True)
-        vmro3 *= o3mulfac
-        vmro3.attrs['units'] = 'nmole mole-1'
+    # o3mulfac = get_unit_conversion_fac("1", vmro3.attrs['units'], 'nmole mole-1')
+    # if o3mulfac != 1:
+    #     vmro3 = vmro3.copy(deep=True)
+    #     vmro3 *= o3mulfac
+    #     vmro3.attrs['units'] = 'nmole mole-1'
```
Comment (avaldebe) [inline pyaerocom/io/_read_mscw_ctm_helpers.py:195]: don't comment old code, remove it

**Review by avaldebe — CHANGES_REQUESTED** [review]
please remove the old commented code, ans answer my question abaout the unused import

File: pyaerocom/io/read_mscw_ctm.py (hunk 3 above)
Comment (dulte) [reply pyaerocom/io/read_mscw_ctm.py:1]: That is an old import I ended up not using. I'm removing it now

**Comment by dulte** [comment]
Old code and import are removed now

**Comment by dulte** [comment]
Saw the rest of the requested changes and they are all implemented now

**Review by avaldebe — APPROVED** [review]
Your changes looks fine to me.
I would like to discuss the tests on i`o.read_mscw_ctm` before you merge this PR.

**Comment by dulte** [comment]
I've now moved the ts typing to it's own method and added an extra test (more a split than an addition)

File: tests/io/test_read_mscw_ctm.py [hunk 23]
```diff
@@ -315,59 +324,283 @@ def test_ReadMscwCtm_aux_var_defs():
...
+                                    vars_and_units=vars_and_units)
+    reader = ReadMscwCtm(data_dir=os.path.join(data_dir, ""))
+    tst = reader.FREQ_CODES[freq]
+    filepaths = reader.filepaths
+    wrong_file = os.path.join(os.path.split(filepaths[0])[0], f"Base_{ts_type}.nc")
+    filepaths[0] = wrong_file
+    new_yrs = reader._get_yrs_from_filepaths()
+
+    
+    cleaned_paths = reader._clean_filepaths(filepaths, new_yrs, tst)
```
Comment (avaldebe) [inline tests/io/test_read_mscw_ctm.py:464]: this test has no asserts

**Review by avaldebe — CHANGES_REQUESTED** [review]
`test_read_emep_wrong_filenames` has no assertions

**Comment by dulte** [comment]
I've added two asserts and changed the name to something more fitting what it actually does

**Pull request merged**